src/vm_cpu.rs

```diff
@@ -95,6 +97,139 @@ pub struct VmCpuTrapState {
     pub htval: u64,
     pub htinst: u64,
 }
+#[derive(Default)]
+#[repr(C)]
+struct VmPmuState {
+    // Stores the mapping for configured PMU counters.
+    configured_counter_map: [u64; drivers::pmu::MAX_COUNTERS],
+    // Stores the values for PMU counters.
+    counter_values: [u64; drivers::pmu::MAX_COUNTERS],
+    // Stores a bit-map of indexes into PmuCounterInfo for configured counters
+    configured_counters: u64,
+    // Stores a bit-map of indexes into PmuCounterInfo for started counters
+    started_counters: u64,
+}
+
+impl VmPmuState {
+    /// Gets the internal state for configured PMU counters.
+    pub fn get_configured_counters(&self) -> u64 {
+        self.configured_counters
+    }
+
+    /// Sets the internal state for started PMU counters.
+    /// This is intended to be called after a successful SBI invocation for starting PMU counters.
+    pub fn set_started_counter(&mut self, counter_index: u64) {
+        if (counter_index as usize) < self.configured_counter_map.len() {
+            self.started_counters |= 1 << counter_index;
+        }
+    }
+
+    /// Gets the internal state for started PMU counters.
+    pub fn get_started_counters(&self) -> u64 {
+        self.started_counters
+    }
+
+    /// Sets the internal state for started PMU counters.
+    /// This is intended to be called after a successful SBI invocation for stopping PMU counters.
+    pub fn set_stopped_counter(&mut self, counter_index: u64) {
+        if (counter_index as usize) < self.configured_counter_map.len() {
+            self.started_counters &= !(1 << counter_index);
+        }
+    }
+
+    /// Sets the internal state for PMU counters.
+    /// This is intended to be called after a successful SBI invocation for configuring PMU counters.
+    pub fn set_counter_configuration(&mut self, counter_index: u64, event_data: u64) {
+        if (counter_index as usize) < self.configured_counter_map.len() {
+            self.configured_counters |= 1 << counter_index;
+            self.configured_counter_map[counter_index as usize] = event_data;
```

Comment:
Where do you clear the mapping in configured_counter_map? It should be cleared in the few places

1. vm reset/destroy
2. SBI_PMU_STOP_FLAG_RESET is set in sbi_pmu_counter_stop 

Did I miss something ?

Reply:
For #1, when the VM is destroyed, do we care about the mappings anymore (since they are specific to the VM)? For the reset case, I believe that they are part of the VM state, and are restored. My impression was that once a counter is configured, it can be stopped and started without reconfiguration, or is that not the case?

Reply:
Once the counter is configured, it can be started/stopped without reconfiguration until reset happens. The perf event flow

perf_event_add
     perf_event_start
     perf_event_stop
perf_event_delete -- this is where reset will happen. This will be called during the context switch of the process.
Thus, another process can do the same flow for some other events.
Once configured it is valid until context switch happens.

Reply:
@atishp04 What is the equivalent of perf_event_delete from the TSM perspective (if any)? I think this is a logical construct used by the perf framework, and from the TSM POV, it will look like yet another call to configure_counter. Does the TSM need to block configuration of already configured counters, or it is legal to re-configure counters?